### PR TITLE
add bastion validator

### PR DESCRIPTION
**Pull request by tedteng**
**How to categorize this PR?**

/area ops-productivity
/kind enhancement

**What this PR does / why we need it**:
This is part of https://github.com/gardener/gardener-extension-provider-alicloud/issues/499 and PR https://github.com/gardener/gardener-extension-provider-alicloud/pull/502/commits/d267cd638974c777e6d471e2c4b629e913691fd2

**Which issue(s) this PR fixes**:
 

**Special notes for your reviewer**:

**Release note**:

```feature dependency
Add Bastion config validator 
```

### CODE REVIEW

**Comment by tedteng** [comment]
after check with @acumino the failed job `pull-gardener-apidiff`  can skip it.

**Comment by petersutter** [comment]
please also adapt https://github.com/gardener/gardener/blob/master/docs/extensions/bastion.md accordingly to https://github.com/gardener/gardener/blob/master/docs/extensions/infrastructure.md#configvalidator-interface

File: extensions/pkg/controller/bastion/configvalidator.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	"context"
+
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/extensions"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// ConfigValidator validates the provider config of bastion resource with the cloud provider.
+type ConfigValidator interface {
+	// Validate validates the provider config of the given infrastructure resource with the cloud provider which used by Bastion.
```
Comment (ary1992) [inline extensions/pkg/controller/bastion/configvalidator.go]:
```suggestion
	// Validate validates the provider config of the given infrastructure resource with the cloud provider used by Bastion.
```

File: extensions/pkg/controller/bastion/controller.go [hunk 2]
```diff
@@ -36,6 +36,8 @@ const (
 type AddArgs struct {
 	// Actuator is a Bastion actuator.
 	Actuator Actuator
+	// ConfigValidator is an bastion config validator.
```
Comment (ary1992) [inline extensions/pkg/controller/bastion/controller.go]:
```suggestion
	// ConfigValidator is a Bastion config validator.
```

**Review by ary1992 — COMMENTED** [review]
Minor Nits

File: extensions/pkg/controller/bastion/configvalidator.go (hunk 1 above)
Comment (petersutter) [reply extensions/pkg/controller/bastion/configvalidator.go]: > of the given infrastructure resource

this is also a copy-paste mistake, bastion and cluster resource is given

Comment (tedteng) [reply extensions/pkg/controller/bastion/configvalidator.go]: commit in 2366c9bda853ed160cf80c46b3f936824656e9eb

Comment (tedteng) [reply extensions/pkg/controller/bastion/configvalidator.go]: commit in 2366c9bda853ed160cf80c46b3f936824656e9eb

File: extensions/pkg/controller/bastion/controller.go (hunk 2 above)
Comment (tedteng) [reply extensions/pkg/controller/bastion/controller.go]: commit in 2366c9bda853ed160cf80c46b3f936824656e9eb

**Comment by rfranzke** [comment]
Can you please add a proper release note?

File: extensions/pkg/controller/bastion/configvalidator.go [hunk 3]
```diff
@@ -24,7 +24,7 @@ import (
 
 // ConfigValidator validates the provider config of bastion resource with the cloud provider.
 type ConfigValidator interface {
-	// Validate validates the provider config of the given infrastructure resource with the cloud provider which used by Bastion.
+	// Validate validates the provider config bastion and cluster resource is given with the cloud provider used by Bastion.
```
Comment (petersutter) [inline extensions/pkg/controller/bastion/configvalidator.go]: Please correct this sentence, it does not make much sense

Comment (tedteng) [reply extensions/pkg/controller/bastion/configvalidator.go]: sorry wrong copy-paste here change back https://github.com/gardener/gardener/commit/3b823f3f48a1660dca9f0020965b0bfdb5709cd2

File: extensions/pkg/controller/bastion/configvalidator.go (hunk 1 above)
Comment (tedteng) [reply extensions/pkg/controller/bastion/configvalidator.go]: I'm not sure if I understand correctly about `bastion and cluster resource is given`. 
currently the this bastion validator purpose only for validator the resource value fetch from instructure status eg vpc name, image id.  these are prerequisites for building bastion objects. need to check it before reconcile the bastion object.  all the bastion resources are still verified in bastion extension logic.

Comment (petersutter) [reply extensions/pkg/controller/bastion/configvalidator.go]: bastion and cluster are passed to the `Validate` function, hence "bastion and cluster resource is given". Previously it said the infrastructure resource is given, which was a copy paste mistake

Comment (tedteng) [reply extensions/pkg/controller/bastion/configvalidator.go]: the object comes from infrastructure eventually, then I use the same comments like infrastructure validator.  now I am understood.  I didn't realize you mention about two objects passed in the Validata function below.   commit in https://github.com/gardener/gardener/pull/6197/commits/ad78cf5608372423e4fb95be9bfb014647fb0855

**Comment by tedteng** [comment]
> Can you please add a proper release note?

updated

**Review by petersutter — APPROVED** [review]
/lgtm

**Review by rfranzke — COMMENTED** [review]
/approve

**Pull request merged**